<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="description" content="">
  <title>Isoform-resolved transcriptome of the human preimplantation embryo</title>
  <script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
  'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','GTM-PF3RRC2');</script>
  <script async src="https://www.googletagmanager.com/gtag/js?id=G-G7VJ4XZS16"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'G-G7VJ4XZS16');
  </script>
  <link rel="icon" type="image/x-icon" href="web-resource/favicon.ico">
  <link href="web-resource/bootstrap-5.2.1/css/bootstrap.min.css" rel="stylesheet">
  <link href="web-resource/style.css" rel="stylesheet">
  <link href="web-resource/fontawesome-free-6.2.0-web/css/fontawesome.css" rel="stylesheet">
  <link href="web-resource/fontawesome-free-6.2.0-web/css/brands.css" rel="stylesheet">
  <link href="web-resource/fontawesome-free-6.2.0-web/css/solid.css" rel="stylesheet">
</head>

<body>
  <noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-PF3RRC2" height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
  
  <main>
    <nav class="navbar navbar-expand-xl mb-3 fixed-top">
      <div class="container">
        <a href="#title" class="d-flex align-items-center my-1 me-md-auto text-dark text-decoration-none scroll-to-top-btn">
          <span class="medium" id="website-title"><b>Isoform-resolved human embryo transcriptome</b></span>
        </a>
    
        <!-- Toggler button for small screens -->
        <button class="navbar-toggler d-none" type="button" data-bs-toggle="collapse" data-bs-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation" style="border: none !important;">
          <!-- <span class="navbar-toggler-icon"></span> -->
          <img src="web-resource/img/rna-menu-2.png" alt="Menu" height="25px">
        </button>
    
        <div class="collapse navbar-collapse justify-content-end mt-2" id="navbarNav">
          <ul class="navbar-nav">
            <li class="nav-item">
              <a class="nav-link active" href="https://github.com/denis-torre/embryo-transcriptome" target="_blank"><img src="web-resource/img/github.png" height="23px"></a>
            </li>
            <li class="nav-item">
              <a class="nav-link active" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190548" target="_blank"><img src="web-resource/img/geo-logo.png" height="23px"></a>
            </li>
            <li class="nav-item">
              <a class="nav-link active" href="https://www.nature.com/articles/s41467-023-42558-y" target="_blank"><img src="web-resource/img/nature-logo.png" height="23px"></a>
            </li>
            <li class="nav-item">
              <a class="nav-link active" href="https://genome.ucsc.edu/s/denis.torre/Embryonic%20transcriptome" target="_blank"><img src="web-resource/img/ucsc-logo.png" height="23px"></a>
            </li>
            <li class="nav-item">
              <a class="nav-link active" href="https://rstudio-connect.hpc.mssm.edu/human-embryo-transcriptome/" target="_blank"><img src="web-resource/img/shiny-logo.png" height="26px"></a>
            </li>
            <!-- <li class="nav-item">
              <a class="nav-link" href="#">About</a>
            </li>
            <li class="nav-item">
              <a class="nav-link" href="#">Services</a>
            </li>
            <li class="nav-item">
              <a class="nav-link" href="#">Contact</a>
            </li> -->
          </ul>
        </div>
      </div>
    </nav>
    <!-- Header -->
    <!-- <div id="title" class="container">
      <header class="d-flex flex-wrap justify-content-center py-3 mb-4 border-bottom">
        <a href="/" class="d-flex align-items-center mb-3 mb-md-0 me-md-auto text-dark text-decoration-none">
          <span class="fs-4 medium">Isoform-resolved human embryo transcriptome</span>
        </a>
      </header>
    </div> -->

    <!-- Content -->
    <div class="container mt-5 pt-5" id="title">

      <!-- Overview -->
      <div class="row">
        <!-- <h3>Overview</h3> -->
        <!-- <p>This website is intended as an accompanying resource to the publication </p> -->
        <div class="col-xl-6 mb-4 mb-xl-0 text-center">
          <img src="web-resource/img/study-design.png" alt="Overview of study design" style="width: 100%;">
        </div>
        <div class="col-xl-6 pt-1">
          <p>
            This website is intended as a resource accompanying the publication <a href="https://www.nature.com/articles/s41467-023-42558-y" target="_blank"><b>Isoform-resolved transcriptome of the human preimplantation embryo</b></a>, published in <i>Nature Communications</i>.
          </p>          
          <p>
            The website contains the following sections:
            <ul>
              <li>
                <b><a href="#genome-browser">Genome browser</a></b>, which contains a genome browser session to explore the data
                <!-- <b><a href="#genome-browser">Genome browser</a></b>, which contains an interactive UCSC genome browser session containing the full isoform-resolved dataset, alongside processed human embryo ATAC-Seq (<a href="https://www.nature.com/articles/s41467-018-08244-0" target="_blank">Liu et al.</a>). and CUT&RUN (<a href="https://www.science.org/doi/10.1126/science.aaw5118" target="_blank">Xia et al.</a>) data, and multiple publicly available tracks (e.g. evolutionary conservation, repetitive element location, ENCODE, GTEx, etc.) -->
              </li>
              <li>
                <b><a href="#isoform-explorer">Isoform explorer</a></b>, which allows users to interactively explore isoforms for genes of interest
              </li>
              <li>
                <b><a href="#data-files">Data files</a></b>, where users can download all relevant processed data files
                <!-- <b><a href="#data-files">Data files</a></b>, where users can download all processed data files containing coordinates, structure and predicted biological properties for each isoform in the embryo transcriptome. -->
              </li>
            </ul>
          </p>          
          Source code used to process the data is fully available on <a href="https://github.com/denis-torre/embryo-transcriptome" target="_blank">GitHub</a>.
        </div>
      </div>
      
      <!-- Data files -->
      <!-- <h3 class="border-bottom pb-3 my-4">Genome browser</h3> -->
      
      
      <!-- Genome browser -->
      <h3 id="genome-browser" class="border-bottom pb-3 my-4">Genome browser <a href="#asd" class="text-black"><i class="fa-solid fa-angle-up scroll-to-top-btn"></i></a></h3>
      The interface below allows users to explore the full isoform-resolved dataset, including RNA-expression across 1C to blastocyst stages; processed human embryo ATAC-Seq (<a href="https://www.nature.com/articles/s41467-018-08244-0" target="_blank">Liu et al.</a>). and CUT&RUN (<a href="https://www.science.org/doi/10.1126/science.aaw5118" target="_blank">Xia et al.</a>) data; and multiple publicly available tracks (e.g. evolutionary conservation, repetitive element location, ENCODE, etc.)
      <br>
      <br>
      The session is also available on the <a href="https://genome.ucsc.edu/s/denis.torre/Embryonic%20transcriptome" target="_blank">UCSC Genome Browser</a>.
      <div class="row my-4">
        <iframe id="genome-browser-iframe" src="https://genome.ucsc.edu/s/denis.torre/Embryonic%20transcriptome?pix=1000" height="2500" width="100%"></iframe>
      </div>

      <!-- Isoform explorer -->
      <h3 id="isoform-explorer" class="border-bottom pb-3 my-4">Isoform explorer <a href="#title" class="text-black"><i class="fa-solid fa-angle-up scroll-to-top-btn"></i></a></h3>
      The interface below allows users to interactively explore isoform-level predictions of biological function for each gene, and view both isoform and gene expression across the samples in our dataset.
      <br>
      <br>
      The isoform explorer is also available at: <a href="https://rstudio-connect.hpc.mssm.edu/human-embryo-transcriptome/" target="_blank">https://rstudio-connect.hpc.mssm.edu/human-embryo-transcriptome/</a>
      <br>
      <br>
      <div class="row my-4">
        <iframe id="isoform-explorer-iframe" class="border border-dark rounded" src="https://denistorre.shinyapps.io/human-embryo-transcriptome/" height="1000" width="100%"></iframe>
      </div>
      

      <!-- Data files -->
      <h3 id="data-files" class="border-bottom pt-2 pb-3 my-4">Data files <a href="#title" class="text-black"><i class="fa-solid fa-angle-up scroll-to-top-btn"></i></a></h3>
      
      <!-- GTF -->
      <div class="row my-4">
        <p>
          Below, users can access all processed long- and short-read RNA-Seq data generated for this study, including isoform-level predictions of biological properties.
        </p>
        <p>
          For a complete explanation of the data, please see our <a href="https://www.nature.com/articles/s41467-023-42558-y#Sec11" target="_blank">Methods</a> and the source code on <a href="https://github.com/denis-torre/embryo-transcriptome" target="_blank">GitHub</a>.
          Raw sequencing data is available on GEO at <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE190548", target="_blank">GSE190548</a>.
        </p>
        <h4 class="mb-3 mt-3">Isoform-resolved transcriptome</h3>
          
        <div class="col-xl-4">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/exon-intron.png" alt="" style="height: 180px;" class="my-2 max-width: 100% !important;">
            </div>
            <b>Isoform coordinates (GTF)</b>
            <p>
              Coordinates of known and novel embryo isoforms on Ensembl GRCh38 genome.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/transcriptome/Human_embryo_isoform_transcriptome_GRCh38.gtf" download>
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Isoform coordinates file (GTF)" data-bs-content="
                  The isoform coordinates file contains information on the <b>genomic coordinates of known and novel isoforms</b> identified in this study.
                  <br><br>
                  This file can be used in place of the standard reference GTF file for <b>transcriptome annotation</b>, including analyses such as gene and isoform-level expression quantification, alternative splicing, and annotation of peaks from ChIP-Seq or ATAC-Seq data.
                  <br><br>
                  The file is in the <b><a href='https://useast.ensembl.org/info/website/upload/gff.html' target='_blank'>GTF format</a></b>, and known isoforms are based on the Ensembl GRCh38 v102 human transcriptome build.
                  " data-bs-html="true">
                  <i class="fa-regular fa-circle-question"></i> Help
                </button>
              </div>
            </div>
          </p>
        </div>
      </div>
      
      <div class="col-xl-4">
        <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
          <div class="text-center">
            <img src="web-resource/img/nucleotide-sequence.png" alt="" style="width: 100%; max-width: 500px !important;" class="py-5">
          </div>
          <b>Isoform sequences (FASTA)</b>
          <p>
            Nucleotide sequences of known and novel embryo isoforms (exons only).
            <br>
            <div class="row text-center">
              <div class="col-6">
                <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/transcriptome/Human_embryo_isoform_transcriptome_GRCh38.fasta">
                  <i class="fa-solid fa-download"></i> Download
                </a>
              </div>
              <div class="col-6">
                <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Isoform sequences file (FASTA)" data-bs-content="
                The isoform sequences file contains information on the <b>nucleotide sequences</b> of the spliced isoforms in the isoform-resolved transcriptome described in this study.
                <br><br>
                This file can be used in place of the standard reference cDNA file for <b>transcript expression quantification</b> using tools such as Kallisto and Salmon. It can also be used to predict <b>isoform biological properties</b> based on their nucleotide sequence, such as protein-coding potential and repetitive element integration.
                <br><br>
                The file is in the <b><a href='https://www.ncbi.nlm.nih.gov/genbank/fastaformat/' target='_blank'>FASTA format</a></b>, and known isoforms are based on the Ensembl GRCh38 v102 human transcriptome build.
                  " data-bs-html="true">
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>

        <div class="col-xl-4">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/isoform-structural-classes.png" alt="" class="py-3" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Isoform structural classes</b>
            <p>
              Isoform classification compared to known transcripts (using TALON/SQANTI3).
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/transcriptome/Human_embryo_isoform_transcriptome_GRCh38-isoform_classification.tsv">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Isoform structural class file (TSV)" data-bs-content="
                  The isoform structural class file contains information on the <b>classification of isoforms</b> in the isoform-resolved transcriptome when compared to known isoforms in the Ensembl GRCh38 v102 human transcriptome build.
                  <br><br>
                  The structural classes of isoforms were predicted using <b><a href='https://github.com/ConesaLab/SQANTI3' target='_blank'>SQANTI3</a></b> (v4.2) and <b><a href='https://github.com/mortazavilab/TALON' target='_blank'>TALON</a></b> (v5.0), and consist of the following:
                  <ul>
                    <li><b>Known</b>: full splice site match to a known transcript</li>
                    <li><b>Novel in catalog (NIC)</b>: novel transcript of a known gene, containing a novel combination of known splice sites</li>
                    <li><b>Novel not in catalog (NNC)</b>: novel transcript of a known gene, containing at least one novel splice site</li>
                    <li><b>Antisense</b>: novel transcript of a novel gene, overlapping a known gene but in antisense direction</li>
                    <li><b>Intergenic</b>: novel transcript of a novel gene, which does not overlap any known gene</li>
                  </ul>
                  For more information on the approach used to generate the transcriptome, please refer to the <b>Methods</b> section of the manuscript. 
                  <br><br>
                  The file is in the TSV format, and contains the following columns:
                  <ul>
                    <li><b>gene_id</b>: Ensembl gene ID (GRCh38 v102 human transcriptome build)</li>
                    <li><b>transcript_id</b>: IDs of known transcripts (beginning with 'ENST') and novel transcripts ('TALONT')</li>
                    <li><b>transcript_length</b>: Length of the transcript primary sequence in nucleotide </li>
                    <li><b>nr_exons</b>: Number of exons in the spliced transcript</li>
                    <li><b>gene_novelty</b>: Gene classification according to TALON</li>
                    <li><b>transcript_novelty</b>: Transcript classification according to TALON</li>
                    <li><b>fl_counts</b>: Full-length PacBio read count </li>
                    <li><b>structural_category</b>: Transcript classification according to SQANTI3</li>
                    <li><b>subcategory</b>: Transcript classification subcategory according to SQANTI3</li>
                    <li><b>gene_novelty_merged</b>: Final gene classification based on integration of TALON and SQANTI3</li>
                    <li><b>Transcript_novelty</b>: Final transcript classification based on integration of TALON and SQANTI3</li>
                  </ul>
                  " data-bs-html="true">
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>

      </div>
      
      <!-- Biological properties -->
      <div class="row my-4">
        <h4 class="mb-3">Isoform biological properties</h3>
          <p> The section below contains links to files with predictes isoform-level biological properties. </p>
        <!-- CPAT -->
        <div class="col-xl-3 col-md-6">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/mrna-translation.png" alt="" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Protein-coding potential</b>
            <p>
              Predicted protein-coding potential and ORFs of each isoform.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/transcriptome_predictions/Human_embryo_isoform_transcriptome_GRCh38-CPAT_ORFs.tsv">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Protein-coding potential file" data-bs-content="
                  The protein-coding potential file contains information on the <b>predicted protein-coding probabilities</b> and <b>best predicted open reading frames</b> (ORFs) of isoforms in the transcriptome.
                  <br><br>
                  These values were predicted using <b><a href='https://cpat.readthedocs.io/en/latest/' target='_blank'>CPAT</a></b> (v3.0.2) based on the isoform nucleotide sequences.
                  <br><br>
                  The file is in the TSV format, and contains the following columns:
                  <ul>
                    <li><b>seq_ID:</b>Transcript ID</li>
                    <li><b>ID:</b>ID of the best predicted ORF for each isoform</li>
                    <li><b>mRNA:</b>Length of the isoform in nucleotides</li>
                    <li><b>ORF_strand:</b>Strand of the best predicted ORF with respect to the isoform</li>
                    <li><b>ORF_frame:</b>Reading frame of the best ORF with respect to the isoform nucleotide sequence</li>
                    <li><b>ORF_start:</b>Start nucleotide of the best ORF with respect to the isoform nucleotide sequence</li>
                    <li><b>ORF_end:</b>End nucleotide of the best ORF with respect to the isoform nucleotide sequence</li>
                    <li><b>ORF:</b>ORF length in nucleotides</li>
                    <li><b>Fickett:</b> <a href='https://academic.oup.com/nar/article/10/17/5303/2385104' target='_blank'>Fickett score</a> of the best predicted ORF (see <a href='https://cpat.readthedocs.io/en/latest/#introduction' target='_blank'>CPAT manual</a> for more details.)</li>
                    <li><b>Hexamer:</b> <a href='https://academic.oup.com/nar/article/20/24/6441/1114215' target='_blank'>Hexamer usage bias</a> of the best predicted ORF (see <a href='https://cpat.readthedocs.io/en/latest/#introduction' target='_blank'>CPAT manual</a> for more details.)</li>
                    <li><b>Coding_prob:</b> predicted protein-coding probability of the ORF. According to the <a href='https://cpat.readthedocs.io/en/latest/#how-to-choose-cutoff' target='_blank'>CPAT manual</a>, the recommended cutoff for protein-coding isoforms in human is â‰¥0.364</li>
                  </ul>                  
                  " data-bs-html="true">
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>
        <!-- Pfam -->
        <div class="col-xl-3 col-md-6">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/protein-domain.png" alt="" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Protein domain content</b>
            <p>
              Predicted protein domains for each predicted protein-coding isoform.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/transcriptome_predictions/Human_embryo_isoform_transcriptome_GRCh38-PfamScan_domains.tsv">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Protein domain content file" data-bs-content="
                  The protein domain content file <b>predicted known protein domains</b> in the translated amino acid sequences of the best ORF of each predicted protein-coding isoform in the isoform-resolved transcriptome.
                  <br><br>
                  These values were predicted using <b>PfamScan</b> (v1.6) and <b>HMMer</b> (v3.3) leveraging the best ORFs predicted using CPAT, as described earlier.
                  <br><br>
                  The file is in the TSV format, and contains the following columns (see <a href='http://eddylab.org/software/hmmer/Userguide.pdf' target='_blank'>HMMer manual</a> for more details):
                  <ul>
                    <li><b>seq_id:</b>Transcript ID</li>
                    <li><b>alignment_start:</b> The position in the target sequence at which the hit starts</li>
                    <li><b>alignment_end:</b> The position in the target sequence at which the hit ends</li>
                    <li><b>envelope_start:</b> The position in the target sequence at which the surrounding envelope starts</li>
                    <li><b>envelope_end:</b> The position in the target sequence at which the surrounding envelope ends</li>
                    <li><b>hmm_acc:</b> The accession of the target HMM (e.g. protein domain)</li>
                    <li><b>hmm_name:</b> The name of the target HMM</li>
                    <li><b>type:</b> The type of the target HMM</li>
                    <li><b>hmm_start:</b> The position in the HMM at which the hit starts</li>
                    <li><b>hmm_end:</b>The position in the HMM at which the hit ends</li>
                    <li><b>hmm_length:</b> The length of the hit in the HMM</li>
                    <li><b>bit_score:</b> The score (in bits) for this hit</li>
                    <li><b>E_value:</b> The expectation value (statistical significance) of the target</li>
                    <li><b>clan:</b> The Pfam 'clan' (a group of two or more families that have arisen from a single evolutionary origin)</li>
                  </ul>
                  " data-bs-html="true">
                  <!-- <li><b>significance:</b></li> -->
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>
        <!-- Repeats -->
        <div class="col-xl-3 col-md-6">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/repetitive-dna.png" alt="" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Repeat element content</b>
            <p>
              Predicted repetitive element content (e.g. ERVs) of each isoform.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/transcriptome_predictions/Human_embryo_isoform_transcriptome_GRCh38-RepeatMasker_scan.tsv">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Repetitive element integration file" data-bs-content="
                  The repetitive element content file contains information on the <b>integration of repetitive elements</b> in the nucleotide sequences of isoforms in the isoform-resolved transcriptome.
                  <br><br>
                  These values were predicted using <b>RepeatMasker</b> (v4.11) using the isoform nucleotide sequences as an input.
                  <br><br>
                  The file is in the TSV format, and contains the following columns (see <a href='https://www.repeatmasker.org/webrepeatmaskerhelp.html' target='_blank'>RepeatMasker manual</a> for more details):
                  <ul>
                    <li><b>sw_score:</b> Smith-Waterman score of the match, usually complexity adjusted</li>
                    <li><b>pct_div:</b> % substitutions in matching region compared to the consensus</li>
                    <li><b>pct_del:</b> % of bases opposite a gap in the query sequence (deleted bp)</li>
                    <li><b>pct_ins:</b> % of bases opposite a gap in the repeat consensus (inserted bp)</li>
                    <li><b>query_sequence:</b> Transcript ID</li>
                    <li><b>query_begin:</b> starting position of match in the transcript</li>
                    <li><b>query_end:</b> ending position of match in the transcript</li>
                    <li><b>query_left:</b> no. of bases in the transcript past the ending position of match</li>
                    <li><b>match:</b> match is with the Complement of the consensus sequence in the database</li>
                    <li><b>matching_repeat:</b> name of the matching interspersed repeat</li>
                    <li><b>repeat_class:</b> the class of the repeat</li>
                    <li><b>repeat_begin:</b> no. of bases in (complement of) the repeat consensus sequence prior to beginning of the match (so 0 means that the match extended all the way to the end of the repeat consensus sequence)</li>
                    <li><b>repeat_end:</b> starting position of match in database sequence (using top-strand numbering)</li>
                    <li><b>repeat_left:</b> ending position of match in database sequence</li>
                  </ul>
                  " data-bs-html="true">
                  <!-- <li><b>ID:</b></li> -->
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>
        <!-- Conservation -->
        <div class="col-xl-3 col-md-6">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/phylogenetic-tree.png" alt="" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Evolutionary conservation</b>
            <p>
              Predicted evolutionary conservation scores for each isoform.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/transcriptome_predictions/Human_embryo_isoform_transcriptome_GRCh38-phastCons100way_average_exon_scores.tsv">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Evolutionary conservation file" data-bs-content="
                  The evolutionary conservation file contains information on the <b>average predicted evolutionary conservation scores</b> of isoforms in the isoform-resolved transcriptome.
                  <br><br>
                  These values were calculated using the <b><a href='https://hgdownload.soe.ucsc.edu/goldenPath/hg38/phastCons100way/' target='_blank'>phastCons100way</a></b> score file, based on the multiple sequence alignment of the human genome to 99 vertebrates spanning primates, mammals, birds, reptiles and fish.
                  <br><br>
                  The file is in the TSV format, and contains the following columns:
                  <ul>
                    <li><b>transcript_id:</b> Transcript ID</li>
                    <li><b>mean_score:</b> Average conservation score across the isoform exons (PhastCons100way)</li>
                  </ul>
                  " data-bs-html="true">
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>
      </div>
      
      <!-- Expression and splicing -->
      <div class="row my-4">
        <h4 class="mb-3">Expression and splicing</h3>
          <p> The section below contains links to files containing estimated expression and splicing of isoforms throughout human preimplantation development. </p>
        <!-- Expression -->
        <div class="col-xl-3 col-md-6">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/expression-heatmap.png" alt="" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Expression counts</b>
            <p>
              Raw and normalized gene and isoform expression levels.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/human_embryo_gene_isoform_expression.zip">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Expression files" data-bs-content="
                  The downloadable ZIP file contains the following files:
                  <ul>
                    <li><b>human_embryo_gene_counts.tsv</b>: raw gene-level counts</li>
                    <li><b>human_embryo_transcript_counts.tsv</b>: raw isoform-level counts</li>
                  </ul>
                  Each file is in the <b>TSV format</b>, with gene or transcript IDs as rows, samples as columns, and estimated counts as values.
                  <br><br>
                  Expression was quantified using STAR (v2.7.5b) and RSEM (v1.3.3).
                  " data-bs-html="true">
                  <!-- <li><b>human_embryo_gene_tpm.tsv</b>: normalized gene-level counts (transcripts per million, TPM)</li> -->
                  <!-- <li><b>human_embryo_transcript_tpm.tsv</b>: normalized isoform-level counts (transcripts per million, TPM)</li> -->
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>
        <!-- DEseq -->
        <div class="col-xl-3 col-md-6">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/volcano-plot.png" alt="" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Differential expression</b>
            <p>
              Differential expression results throughout development.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/human_embryo_differential_expression.zip">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <!-- <li><b>human_embryo_differential_transcript_expression.xlsx</b>: isoform-level differential expression</li> -->
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Differential expression files" data-bs-content="
                  The downloadable ZIP file contains the following files:
                  <ul>
                    <li><b>human_embryo_differential_gene_expression.xlsx</b>: gene-level differential expression</li>
                  </ul>
                  Each file is in the <b>XLSX format</b>, with each sheet containing a pairwise comparison of consecutive preimplantation developmental stages.
                  <br><br>
                  Each sheet contains the following columns (default output from DESeq2, v1.3.0):
                  <ul>
                    <li><b>gene_id</b>: gene ID (known genes begin with 'ENSG', novel genes begin with 'TALONG')</li>
                    <li><b>transcript_id</b> (only for isoform-level results file): transcript ID (known transcripts begin with 'ENST', novel genes begin with 'TALONT')</li>
                    <li><b>baseMean</b>: average normalized expression across samples</li>
                    <li><b>log2FoldChange</b>: log2-transformed fold change in normalized expression across stages. Positive values indicate higher expression in the later developmental stage, while negative values indicate lower expression in the later developmental stage.</li>
                    <li><b>pvalue</b>: unadjusted differential expression p-value from DESeq2.</li>
                    <li><b>padj</b>: adjusted p-value using the Benjamini-Hochberg method.</li>
                  </ul>
                  " data-bs-html="true">
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>
        <!-- PSI -->
        <div class="col-xl-3 col-md-6">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/psi-heatmap.png" alt="" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Splicing inclusion levels</b>
            <p>
              Relative isoform inclusion levels across samples.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://torred23.dmz.hpc.mssm.edu/human-embryo-transcriptome/human_embryo_splicing_psi.zip">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Splicing inclusion files" data-bs-content="
                  The downloadable ZIP file contains the following files, each containing percent spliced in (PSI) levels across embryo samples:
                  <ul>
                    <li><b>human_embryo_transcript_psi.tsv</b>: isoform-level</li>
                    <li><b>human_embryo_SE_psi.tsv</b>: skipped exon (SE)</li>
                    <li><b>human_embryo_RI_psi.tsv</b>: retained intron (RI)</li>
                    <li><b>human_embryo_A5_psi.tsv</b>: alternative 5' splice site (A5)</li>
                    <li><b>human_embryo_A3_psi.tsv</b>: alternative 3' splice site (A3)</li>
                    <li><b>human_embryo_MX_psi.tsv</b>: mutually exclusive exons (MX)</li>
                    <li><b>human_embryo_AF_psi.tsv</b>: alternative first exon (AF)</li>
                    <li><b>human_embryo_AL_psi.tsv</b>: alternative last exon (AL)</li>
                  </ul>
                  Each file is in the <b>TSV format</b>, containing transcript or splicing event IDs on rows, samples on columns, and PSI as values.
                  <br><br>
                  Results were generated using <b>SUPPA2</b> (v2.3). For more information on the data and format, please see the <a href='https://github.com/comprna/SUPPA' target='_blank'>SUPPA manual</a>.
                  " data-bs-html="true">
                  <!-- <li><b>ID:</b></li> -->
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>
        <!-- dPSI -->
        <div class="col-xl-3 col-md-6">
          <div class="border rounded px-3 py-2 mx-1 my-2 my-xl-0 height-100">
            <div class="text-center">
              <img src="web-resource/img/alternative-splicing.png" alt="" style="height: 200px; max-width: 100% !important;">
            </div>
            <b>Alternative splicing</b>
            <p>
              Alternative splicing events throughout development.
              <br>
              <div class="row text-center">
                <div class="col-6">
                  <a class="btn btn-light w-100" href="https://human-embryo-transcriptome.s3.amazonaws.com/human_embryo_alternative_splicing.zip">
                    <i class="fa-solid fa-download"></i> Download
                  </a>
                </div>
                <div class="col-6">
                  <button type="button" class="btn btn-light w-100" data-bs-container="body" data-bs-toggle="popover" data-bs-placement="right" data-bs-title="Alternative splicing files" data-bs-content="
                  The downloadable ZIP file contains the following files, each containing alternative splicing results across developmental stage transitions:
                  <ul>
                    <li><b>human_embryo_alternative_isoform_inclusion.xlsx</b>: isoform-level results</li>
                    <li><b>human_embryo_alternative_splicing_events.xlsx</b>: splicing-event level results</li>
                  </ul>
                  Each file is in the <b>XLSX format</b>, with each sheet containing a pairwise comparison of consecutive preimplantation developmental stages.
                  <br><br>
                  Each sheet contains the following columns (default output from SUPPA2, v2.3):
                  <ul>
                    <li><b>event_id</b>: ID of the splicing event (see <a href='https://github.com/comprna/SUPPA' target='_blank'>SUPPA manual</a> for details)</li>
                    <li><b>dPSI</b>: differential percent spliced in (PSI) across compared developmental stages</li>
                    <li><b>avg_logTPM</b>: average log-transformed transcripts per million (TPM) of the splicing event across samples</li>
                    <li><b>pvalue</b>: alternative splicing p-value computed by SUPPA2</li>
                    <li><b>event_type</b>: type of splicing event (see <a href='https://github.com/comprna/SUPPA' target='_blank'>SUPPA manual</a> for details)</li>
                  </ul>
                  Results were generated using <b>SUPPA2</b> (v2.3). For more information on the data and format, please see the <a href='https://github.com/comprna/SUPPA' target='_blank'>SUPPA manual</a>.
                  " data-bs-html="true">
                    <i class="fa-regular fa-circle-question"></i> Help
                  </button>
                </div>
              </div>
            </p>
          </div>
        </div>
      </div>
      Icons generated using <a href="https://biorender.com/" target="_blank">BioRender.com</a>.<br>

      <!-- Pipeline -->
      <!-- <div class="row my-4">
        <h4 class="mb-3">Source code</h3>
          <div class="col-5">
            <img src="web-resource/img/analysis-pipeline.png" alt="" style="width: 100%">
          </div>
        <div class="col-7">
          <ul>
            Link to <a href="https://github.com/denis-torre/early-embryo" target="_blank">GitHub</a>
          </ul>
        </div>
      </div> -->

      <!-- Acknowledgements -->
      <div class="row my-4" id="acknowledgements">
        <h4 class="border-bottom pb-3 my-4">Acknowledgements</h3>
        <div class="col-lg-6">
          <div class="row">
            <div class="col-lg-4">
              <a href="" target="_blank">
                <img src="web-resource/img/ismms.jpg" alt="" style="width: 100%; max-width: 300px;" class="mb-3 mb-md-0">
              </a>
            </div>
            <div class="col-lg-8 pb-5">
              <a href="https://icahn.mssm.edu/research/advanced-genomics-technology" target="_blank">Sebra lab<br>Center for Advanced Genomics Technologies (CAGT)</a>
              <br>
              <a href="https://icahn.mssm.edu/research/genomics" target="_blank">Department of Genetics and Genomic Sciences</a>
            </div>
          </div>
        </div>
        <div class="col-lg-6">
          <div class="row">
            <div class="col-lg-2">
              <a href="" target="_blank">
                <img src="web-resource/img/tasmc.png" alt="" style="max-width: 100%; max-height: 100px;" class="mb-3 mb-md-0">
              </a>
            </div>
            <div class="col-lg-10">
              <a href="https://rnd.tasmc.org.il/laboratories/the-wolfe-pgd-stem-cell-research-lab-in-the-institution-of-reproduction-and-ivf-prof-dalit-ben-yosef/" target="_blank">Wolfe PGD-Stem Cell Research Lab</a>
              <br>
              <a href="https://www.tasmc.org.il/sites/en/Lis/Fertility-Research/Pages/Lab-IVF.aspx" target="_blank">Tel Aviv Sourasky Medical Center IVF Laboratory</a>
              <br>
              <a href="https://en-sagol.tau.ac.il/researchers/Dalit-Ben-Yosef" target="_blank">Sagol School of Neuroscience, Tel Aviv University</a>
            </div>
          </div>
        </div>
        </div>
      
    </div>


  </main>


  <script src="web-resource/bootstrap-5.2.1/js/bootstrap.bundle.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/javascript.util/0.12.12/javascript.util.min.js" integrity="sha512-oHBLR38hkpOtf4dW75gdfO7VhEKg2fsitvHZYHZjObc4BPKou2PGenyxA5ZJ8CCqWytBx5wpiSqwVEBy84b7tw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
  <script type="text/javascript" src="https://livejs.com/live.js"></script>
  <script src="https://code.jquery.com/jquery-3.2.1.min.js" crossorigin="anonymous"></script>
  <script type="text/javascript">
    const popoverTriggerList = document.querySelectorAll('[data-bs-toggle="popover"]')
    const popoverList = [...popoverTriggerList].map(popoverTriggerEl => new bootstrap.Popover(popoverTriggerEl))
    // Popover dismiss
    $('body').on('click', function (e) {
      if ($(e.target).parents('.popover').length==0) {
      (($('[data-bs-toggle="popover"]').popover('hide').data('bs.popover') || {}).inState || {}).click = false;
    }
  });
    $('.scroll-to-top-btn').on('click', function(event) {
      event.preventDefault();

      // Use jQuery's animate function to smoothly scroll to the top of the page
      $('html, body').animate({
        scrollTop: 0
      }, 100); // You can adjust the animation speed (in milliseconds) as needed
    });  </script>
</body>
</html>